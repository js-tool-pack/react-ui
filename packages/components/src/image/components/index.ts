export * from './Preview';
export * from './Preview.Toolbar';
export * from './Preview.Group';
