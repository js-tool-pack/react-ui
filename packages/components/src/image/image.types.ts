import { PropsBase } from '@pkg/shared';
import React from 'react';

type ImgAttrs = Partial<React.ImgHTMLAttributes<HTMLImageElement>>;
type ImgStyles = Partial<React.CSSProperties>;

export interface ImageProps
  extends Omit<PropsBase<HTMLDivElement>, 'children'>,
    Pick<Partial<HTMLImageElement>, 'height' | 'width' | 'src' | 'alt'> {
  fit?: ImgStyles['objectFit'];
  imgAttrs?: ImgAttrs;
  name?: string;
}
