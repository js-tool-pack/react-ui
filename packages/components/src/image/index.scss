@use '../namespace' as Name;

$r: Name.$image;
$p: Name.$image-preview;
$t: #{Name.$image-preview}-toolbar;

.#{$r} {
  display: inline-flex;
  &--preview {
    .#{$r}__img {
      cursor: pointer;
    }
  }
}

.#{$p} {
  position: fixed;
  z-index: 1000;
  display: flex;
  justify-content: center;
  align-items: center;
  inset: 0;
  background-color: var(--t-mask-bg-color-reverse);
  &__img {
    cursor: grab;
    max-width: 100%;
    max-height: 100%;
    transition: all var(--t-transition-duration);
  }
  &__toolbar-wrapper {
    position: absolute;
    right: 0;
    bottom: 20px;
    left: 0;
    display: flex;
    align-items: center;
    margin: auto;
    flex-direction: column;
  }
  &__progress {
    margin-bottom: 20px;
    color: white;
  }
}
.#{$t} {
  $h: 48px;
  $p: 12px;

  display: flex;
  align-items: center;
  padding: 0 $p;
  border-radius: calc($h / 2);
  line-height: $h;
  background-color: rgb(0 0 0 / 35%);
  box-sizing: content-box;
  > div {
    display: flex;
    justify-content: center;
    align-items: center;
    width: $h;
    height: $h;
    cursor: pointer;
    color: #d0d0d0;
    &:hover {
      color: white;
    }
  }
}
