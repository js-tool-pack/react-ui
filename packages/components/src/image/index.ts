export type { ImageProps } from './image.types';
export * from './Image';
