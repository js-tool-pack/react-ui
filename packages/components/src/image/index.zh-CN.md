---
category: Components
title: Image 图像
atomId: Image
demo:
  cols: 2
group:
  title: 通用
---

Image 图像。

## 代码演示

<!-- prettier-ignore -->
<code src="./demo/basic.tsx"></code>

## API

Image 的属性说明如下：

| 属性  | 说明          | 类型                                            | 默认值 | 版本 |
| ----- | ------------- | ----------------------------------------------- | ------ | ---- |
| --    | --            | --                                              | --     | --   |
| attrs | html 标签属性 | Partial\<React.HTMLAttributes\<HTMLDivElement>> | --     | --   |

其他说明。
