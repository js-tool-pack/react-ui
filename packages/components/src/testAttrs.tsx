import { fireEvent, render } from '@testing-library/react';
import type { PropsBase } from '@pkg/shared';
import React from 'react';

export function testAttrs<
  T extends { attrs?: Partial<React.HTMLAttributes<HTMLElement>> },
>(component: React.FC<T>, displayName?: string) {
  const name = displayName || component.displayName || component.name;
  test(name ? `${name} attrs` : `attrs`, () => {
    const onClick = jest.fn();

    const FC = component as React.FC<PropsBase>;
    const { container } = render(
      <FC
        attrs={{
          style: { width: '20px' },
          className: 'foo',
          role: 'note',
          onClick,
        }}
      />,
    );

    expect(container.firstChild).toHaveClass('foo');
    expect(container.firstChild).toHaveAttribute('role', 'note');
    expect(container.firstChild).toHaveStyle({ width: '20px' });

    expect(onClick).not.toBeCalled();
    fireEvent.click(container.firstChild!);
    expect(onClick).toBeCalled();
  });
}
