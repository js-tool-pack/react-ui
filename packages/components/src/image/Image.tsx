import React, { useContext, useEffect, useState, useRef } from 'react';
import { ImagePreviewGroupContext } from '~/image/components';
import { useForceUpdate, getClasses } from '@pkg/shared';
import type { RequiredPart } from '@tool-pack/types';
import { getClassNames } from '@tool-pack/basic';
import type { ImageProps } from './image.types';
import { ImagePreview } from '@pkg/components';

const cls = getClasses('image', ['img', 'fallback'], ['preview']);
const defaultProps = {
  fallback:
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMIAAADDCAYAAADQvc6UAAABRWlDQ1BJQ0MgUHJvZmlsZQAAKJFjYGASSSwoyGFhYGDIzSspCnJ3UoiIjFJgf8LAwSDCIMogwMCcmFxc4BgQ4ANUwgCjUcG3awyMIPqyLsis7PPOq3QdDFcvjV3jOD1boQVTPQrgSkktTgbSf4A4LbmgqISBgTEFyFYuLykAsTuAbJEioKOA7DkgdjqEvQHEToKwj4DVhAQ5A9k3gGyB5IxEoBmML4BsnSQk8XQkNtReEOBxcfXxUQg1Mjc0dyHgXNJBSWpFCYh2zi+oLMpMzyhRcASGUqqCZ16yno6CkYGRAQMDKMwhqj/fAIcloxgHQqxAjIHBEugw5sUIsSQpBobtQPdLciLEVJYzMPBHMDBsayhILEqEO4DxG0txmrERhM29nYGBddr//5/DGRjYNRkY/l7////39v///y4Dmn+LgeHANwDrkl1AuO+pmgAAADhlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAAqACAAQAAAABAAAAwqADAAQAAAABAAAAwwAAAAD9b/HnAAAHlklEQVR4Ae3dP3PTWBSGcbGzM6GCKqlIBRV0dHRJFarQ0eUT8LH4BnRU0NHR0UEFVdIlFRV7TzRksomPY8uykTk/zewQfKw/9znv4yvJynLv4uLiV2dBoDiBf4qP3/ARuCRABEFAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghggQAQZQKAnYEaQBAQaASKIAQJEkAEEegJmBElAoBEgghgg0Aj8i0JO4OzsrPv69Wv+hi2qPHr0qNvf39+iI97soRIh4f3z58/u7du3SXX7Xt7Z2enevHmzfQe+oSN2apSAPj09TSrb+XKI/f379+08+A0cNRE2ANkupk+ACNPvkSPcAAEibACyXUyfABGm3yNHuAECRNgAZLuYPgEirKlHu7u7XdyytGwHAd8jjNyng4OD7vnz51dbPT8/7z58+NB9+/bt6jU/TI+AGWHEnrx48eJ/EsSmHzx40L18+fLyzxF3ZVMjEyDCiEDjMYZZS5wiPXnyZFbJaxMhQIQRGzHvWR7XCyOCXsOmiDAi1HmPMMQjDpbpEiDCiL358eNHurW/5SnWdIBbXiDCiA38/Pnzrce2YyZ4//59F3ePLNMl4PbpiL2J0L979+7yDtHDhw8vtzzvdGnEXdvUigSIsCLAWavHp/+qM0BcXMd/q25n1vF57TYBp0a3mUzilePj4+7k5KSLb6gt6ydAhPUzXnoPR0dHl79WGTNCfBnn1uvSCJdegQhLI1vvCk+fPu2ePXt2tZOYEV6/fn31dz+shwAR1sP1cqvLntbEN9MxA9xcYjsxS1jWR4AIa2Ibzx0tc44fYX/16lV6NDFLXH+YL32jwiACRBiEbf5KcXoTIsQSpzXx4N28Ja4BQoK7rgXiydbHjx/P25TaQAJEGAguWy0+2Q8PD6/Ki4R8EVl+bzBOnZY95fq9rj9zAkTI2SxdidBHqG9+skdw43borCXO/ZcJdraPWdv22uIEiLA4q7nvvCug8WTqzQveOH26fodo7g6uFe/a17W3+nFBAkRYENRdb1vkkz1CH9cPsVy/jrhr27PqMYvENYNlHAIesRiBYwRy0V+8iXP8+/fvX11Mr7L7ECueb/r48eMqm7FuI2BGWDEG8cm+7G3NEOfmdcTQw4h9/55lhm7DekRYKQPZF2ArbXTAyu4kDYB2YxUzwg0gi/41ztHnfQG26HbGel/crVrm7tNY+/1btkOEAZ2M05r4FB7r9GbAIdxaZYrHdOsgJ/wCEQY0J74TmOKnbxxT9n3FgGGWWsVdowHtjt9Nnvf7yQM2aZU/TIAIAxrw6dOnAWtZZcoEnBpNuTuObWMEiLAx1HY0ZQJEmHJ3HNvGCBBhY6jtaMoEiJB0Z29vL6ls58vxPcO8/zfrdo5qvKO+d3Fx8Wu8zf1dW4p/cPzLly/dtv9Ts/EbcvGAHhHyfBIhZ6NSiIBTo0LNNtScABFyNiqFCBChULMNNSdAhJyNSiECRCjUbEPNCRAhZ6NSiAARCjXbUHMCRMjZqBQiQIRCzTbUnAARcjYqhQgQoVCzDTUnQIScjUohAkQo1GxDzQkQIWejUogAEQo121BzAkTI2agUIkCEQs021JwAEXI2KoUIEKFQsw01J0CEnI1KIQJEKNRsQ80JECFno1KIABEKNdtQcwJEyNmoFCJAhELNNtScABFyNiqFCBChULMNNSdAhJyNSiECRCjUbEPNCRAhZ6NSiAARCjXbUHMCRMjZqBQiQIRCzTbUnAARcjYqhQgQoVCzDTUnQIScjUohAkQo1GxDzQkQIWejUogAEQo121BzAkTI2agUIkCEQs021JwAEXI2KoUIEKFQsw01J0CEnI1KIQJEKNRsQ80JECFno1KIABEKNdtQcwJEyNmoFCJAhELNNtScABFyNiqFCBChULMNNSdAhJyNSiECRCjUbEPNCRAhZ6NSiAARCjXbUHMCRMjZqBQiQIRCzTbUnAARcjYqhQgQoVCzDTUnQIScjUohAkQo1GxDzQkQIWejUogAEQo121BzAkTI2agUIkCEQs021JwAEXI2KoUIEKFQsw01J0CEnI1KIQJEKNRsQ80JECFno1KIABEKNdtQcwJEyNmoFCJAhELNNtScABFyNiqFCBChULMNNSdAhJyNSiEC/wGgKKC4YMA4TAAAAABJRU5ErkJggg==',
  preview: true,
} satisfies Partial<ImageProps>;

export const Image: React.FC<ImageProps> = React.forwardRef<
  HTMLDivElement,
  ImageProps
>((props, ref) => {
  const {
    imgAttrs = {},
    attrs = {},
    fallback,
    preview,
    height,
    width,
    lazy,
    src,
    fit,
    alt,
  } = props as RequiredPart<ImageProps, keyof typeof defaultProps>;

  const context = useContext(ImagePreviewGroupContext);

  useEffect(() => {
    if (!context || !preview) return;
    const url = src || imgAttrs.src || '';
    if (context.includes(url)) return;
    context.push(url);
    return () => {
      const index = context.indexOf(url);
      if (index === -1) return;
      context.splice(index, 1);
    };
  }, [preview]);

  const [imgVisible, setImgVisible] = useState(true);
  const [previewVisible, setPreviewVisible] = useState(false);
  const imgRef = useRef<HTMLImageElement>(null);
  const forceUpdate = useForceUpdate();
  const lazyingRef = useRef<boolean | void>(undefined);

  if (lazyingRef.current === undefined && lazy) {
    lazyingRef.current = true;
  }

  useEffect(() => {
    const el = imgRef.current;
    if (!lazy) return;
    if (!el || !lazyingRef.current) return;

    const handler = (entries: IntersectionObserverEntry[]) => {
      if (entries[0]!.intersectionRatio <= 0) return;
      lazyingRef.current = false;
      forceUpdate();
    };
    const observer = new IntersectionObserver(handler);
    observer.observe(el);
    return () => observer.disconnect();
  }, [lazy]);

  const Img = (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-noninteractive-element-interactions
    <img
      onClick={(e) => {
        if (lazyingRef.current) return;
        imgAttrs.onClick?.(e);
        setPreviewVisible(true);
      }}
      style={{
        ...imgAttrs.style,
        objectFit: fit || imgAttrs.style?.objectFit,
      }}
      src={lazyingRef.current ? fallback : src || imgAttrs.src}
      height={height || imgAttrs.height}
      width={width || imgAttrs.width}
      alt={alt || imgAttrs.alt}
      className={cls.__.img}
      onError={hideImg}
      onAbort={hideImg}
      ref={imgRef}
      key={1}
    />
  );

  const Fallback = (
    <img
      style={{
        ...imgAttrs.style,
        objectFit: fit || imgAttrs.style?.objectFit,
      }}
      height={height || imgAttrs.height}
      width={width || imgAttrs.width}
      className={cls.__.fallback}
      alt={alt || imgAttrs.alt}
      src={fallback}
    />
  );

  return (
    <div
      {...attrs}
      className={getClassNames(cls.root, attrs.className, {
        [cls['--'].preview]: preview,
      })}
      ref={ref}
    >
      {imgVisible
        ? [
            Img,
            preview && previewVisible && (
              <ImagePreview
                onHide={() => setPreviewVisible(false)}
                images={[src]}
                key={2}
              />
            ),
          ]
        : Fallback}
    </div>
  );

  function hideImg(): void {
    setImgVisible(false);
  }
});

Image.defaultProps = defaultProps;
Image.displayName = 'Image';
