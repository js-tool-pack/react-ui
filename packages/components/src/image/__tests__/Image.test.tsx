import { fireEvent, render } from '@testing-library/react';
import { ImagePreviewGroup, Image } from '~/image';
import { testAttrs } from '~/testAttrs';

describe('Image', () => {
  testAttrs(Image);

  test('basic', () => {
    render(<Image />);
    expect(document.body).toMatchSnapshot();
  });

  describe('preview', () => {
    test('disabled', () => {
      render(<Image preview={false} />);
      expect(document.body).toMatchSnapshot();
    });

    test('enabled', () => {
      render(<Image />);
      expect(document.body).toMatchSnapshot();
    });
  });

  test('group', () => {
    render(
      <ImagePreviewGroup>
        <Image src="https://test.com/a.png" />
        <Image src="https://test.com/b.png" />
        <Image src="https://test.com/c.png" />
      </ImagePreviewGroup>,
    );

    fireEvent.click(document.querySelectorAll('.t-image__img')![1]!);

    expect(
      document.querySelector<HTMLImageElement>('.t-image-preview img')!.src,
    ).toBe('https://test.com/b.png');
    expect(
      document.querySelector('.t-image-preview__progress'),
    ).toHaveTextContent('2 / 3');
  });
});
