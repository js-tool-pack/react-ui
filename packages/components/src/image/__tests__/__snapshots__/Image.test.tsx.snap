// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image basic 1`] = `
<body>
  <div>
    <div
      class="t-image t-image--preview"
    >
      <img
        class="t-image__img"
      />
    </div>
  </div>
</body>
`;

exports[`Image preview disabled 1`] = `
<body>
  <div>
    <div
      class="t-image"
    >
      <img
        class="t-image__img"
      />
    </div>
  </div>
</body>
`;

exports[`Image preview enabled 1`] = `
<body>
  <div>
    <div
      class="t-image t-image--preview"
    >
      <img
        class="t-image__img"
      />
    </div>
  </div>
</body>
`;
